lib/Exporter/ONNXModelWriter.cpp

```diff
@@ -0,0 +1,1367 @@
+    // Otherwise call common operators method or special operators and write
+    // protobuf inputs from node inputs and protobuf outputs from uses.
+    GraphPostOrderVisitor visitor(G_);
+    for (const auto *N : visitor.getPostOrder()) {
+      if (isIgnoreNode(N)) {
+        // Skip ignored nodes.
+        continue;
+      }
+
+      if (reportedNodes.count(N)) {
+        continue;
+      }
+
+      const auto kind = N->getKind();
+      if (kind == Kinded::Kind::PlaceholderKind ||
+          kind == Kinded::Kind::ConstantKind) {
+        if (hasUsesOfKind(N, Kinded::Kind::SaveNodeKind)) {
+          // Storage as an input to SaveNode - ignore it.
+          continue;
+        }
+
+        // Handle placeholders cases.
+        if (kind == Kinded::Kind::PlaceholderKind) {
+          const auto *PH = llvm::dyn_cast<Placeholder>(N);
+          // Write global input, output only tensor shape.
+          auto *inputProto = graphProto->add_input();
+          tensorShapeFromPlaceholder(PH, inputProto);
+        } else {
+          // Write global initializer, output tensor bytes.
+          const auto *C = llvm::dyn_cast<Constant>(N);
+          auto *tensorProto = graphProto->add_initializer();
+          tensorProto->set_name(C->getName());
+          writeTensor(C->getPayload(), tensorProto);
+        }
+      } else if (kind == Kinded::Kind::SaveNodeKind) {
+        // Save node case, find input and use its name as a global output,
+        // output only shape.
+        const auto *PH = llvm::dyn_cast<SaveNode>(N)->getPlaceholder();
+        auto *outputProto = graphProto->add_output();
+        tensorShapeFromPlaceholder(PH, outputProto);
+      } else {
+        RETURN_IF_ERR(writeOperator(N, *graphProto, reporter));
+      }
+      reporter(N);
+    }
+    return writeModel(modelProto, textMode);
+  };
+
+  if (errPtr) {
+    *errPtr = setup();
+  } else {
+    EXIT_ON_ERR(setup());
+  }
+}
+
+ONNXModelWriter::TensorType::DataType
+ONNXModelWriter::convertType(const Type &glowType) {
+  switch (glowType.getElementType()) {
+  case ElemKind::FloatTy:
+    return TensorType::FLOAT;
+  case ElemKind::Float16Ty:
+    return TensorType::FLOAT16;
+  case ElemKind::Int8QTy:
+    return TensorType::INT8;
+  case ElemKind::UInt8FusedQTy:
+  case ElemKind::UInt8QTy:
+    return TensorType::UINT8;
+  case ElemKind::Int16QTy:
+    return TensorType::INT16;
+  case ElemKind::Int32QTy:
+  case ElemKind::Int32ITy:
+    return TensorType::INT32;
+  case ElemKind::Int64ITy:
+    return TensorType::INT64;
+  case ElemKind::BoolTy:
+    return TensorType::BOOL;
+  }
+}
+
+void ONNXModelWriter::writeTensor(const Tensor &T, TensorType *out) {
+  const auto &type = T.getType();
+  out->set_data_type(convertType(type));
+  const auto &dims = type.dims();
+  for (unsigned b = 0, e = dims.size(); b < e; ++b) {
+    out->add_dims(dims[b]);
+  }
+
+  out->set_raw_data(std::string(T.getUnsafePtr(), type.getSizeInBytes()));
```

Comment:
Question: Do we need to care about the endianness of the tensor data? Does ONNX prescribe what needs to be used?

Reply:
I think default is little-endian:
https://github.com/Microsoft/onnxruntime/blob/master/onnxruntime/core/framework/tensorprotoutils.cc#L49

Reply:
OK. It means we need to take care of always producing little-endian here.